### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by czerasz**
Build the application binary inside a Docker container.

Simply run:

    ./scripts/build

This creates the application binary in the project directory:

    $ ls -alh app
    -rwxr-xr-x 1 user user 18M Mai 10 17:06 app
    $ ./app --version
    app version 0.0.3-20-gd6e386a

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+# Stop on any error
+set -e
+
+script_dirirectory="$( cd "$( dirname "$0" )" && pwd )"
```
Comment (purpleidea) [inline scripts/build]: Should these be directory instead of dirirectory ?

File: scripts/build [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+docker rm -f mgmt-export 2> /dev/null
+
+# Start the container in background so we can "copy out" the binary
+docker run -d --name=mgmt-export $image_name bash -c 'while true; sleep 1000; done'
+
+# Remove the current binary
+rm $project_dirirectory/app
+
+# Get the binary from the container
+docker cp mgmt-export:/go/src/app/app $project_dirirectory/app
```
Comment (purpleidea) [inline scripts/build]: Should we name it "mgmt" instead of app? Why or why not?

File: scripts/build [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+
+# Stop on any error
+set -e
+
+script_dirirectory="$( cd "$( dirname "$0" )" && pwd )"
+project_dirirectory=$script_dirirectory/..
+
+# Specify the Docker image name
+image_name='mgmt/mgmt'
```
Comment (purpleidea) [inline scripts/build]: Would it be better if this was purpleidea/mgmt ?

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
+FROM golang:1.6.2
+
+MAINTAINER Michał Czeraszkiewicz <contact@czerasz.com>
+
+# Set the reset cache variable
+ENV REFRESHED_AT 2016-05-10
```
Comment (purpleidea) [inline docker/Dockerfile]: Can you explain this? I'm actually not familiar with it. Thanks!

**Comment by purpleidea** [comment]
Hi @czerasz and thanks for the patch. I left you a few comments for review, and lastly, but more importantly, I'm happy to merge this, but in particular if it's something you think will be useful and that you plan on consuming via this method, I'd love to get a bit more background. In particular it would be great if you plan to keep this portion maintained at least for the imminent future.

Thanks!

File: scripts/build (hunk 2 above)
Comment (czerasz) [inline scripts/build]: When I run `make build` it creates the `app` file. I can rename it to `mgmt`. What do you think?

If you mean the directory inside the container (`/go/src/app/app`), then it really doesn't make much difference how you call it. I took the name from [here](https://github.com/docker-library/golang/blob/ce284e14cdee73fbaa8fb680011a812f272eae2e/1.6/onbuild/Dockerfile#L4).

File: scripts/build (hunk 3 above)
Comment (czerasz) [inline scripts/build]: It's o for me to rename it to `purpleidea/mgmt`. I would ask you to create an automated build from this repo.

File: docker/Dockerfile (hunk 4 above)
Comment (czerasz) [inline docker/Dockerfile]: Find more details [here](http://czerasz.com/2014/11/13/docker-tip-and-tricks/#use-refreshedat-variable-for-better-cache-control).

**Comment by czerasz** [comment]
@purpleidea I don't know much about `mgmt` yet... Not sure how useful this will be in the future...? But since I don't like installing software directly on my machine anymore and since it might be useful for [CoreOS](https://coreos.com/) I thought it's a cool idea to run `mgmt` it inside a container.

Something even more cool would be the possibility to [develop inside a container](https://github.com/czerasz/mgmt/commit/75ed9238b93f3fc9e8cfede8fbb3df8eba69494e).

If you agree it's useful (developing inside a container and distributing `mgmt` inside the container) then I can take responsibility for that.

**Comment by purpleidea** [comment]
> When I run make build it creates the app file. I can rename it to mgmt. What do you think?

I think it's a good idea to rename, so that when you run mgmt (which uses the PROGRAM) variable name, it still shows up as mgmt. If you don't agree about the reasoning, I'm happy to defer to you on this one.

Other than that, the patches look ready to merge. I'll wait until you LMK RE: the above. Also, please rebase the last three patches into your main docker patches. I like your idea about being the docker maintainer, and I am glad to have you participating in the project!

Cheers!

**Comment by purpleidea** [comment]
@czerasz Got a question on IRC, but then you dissapeared offline. RE: your infra questions, etcd will be packaged into the one binary. This works, but is not yet merged in git master yet.

**Comment by purpleidea** [comment]
@czerasz RE: Review, did you also change the binary name to mgmt in the Dockerfile? IE: is this currently correct?

In any case, all of this needs a merge and a rebase, there's no reason there should be a merge commit for example. If you don't know how to do this, let us know and we or #git IRC can help.

**Comment by purpleidea** [comment]
@czerasz Okay this is looking pretty good... Some small fixes:

1) Can you put all of the scripts/ docker stuff inside the docker/ folder too please?

2) Please remove all the leading whitespace in the .sh files you created. Just put the whole commands on oneline, or indent with a tab, not spacespacespace.

3) When you make these two changes, do git add <files you updated after fixing 1 and 2>. And then do `git commit --amend` to put those changes in this original commit!!

4) Rebase to git master. (git checkout master, git pull, git checkout <thisbranch>, git rebase master, git push <yourremote> <thisbranch> --force)

Looks like the home stretch!

**Comment by purpleidea** [comment]
@czerasz Still hacking on this one?

**Comment by czerasz** [comment]
@purpleidea it's done. Sorry for the delay.

**Comment by purpleidea** [comment]
@czerasz Please don't apologize, and welcome to the contributors circle! Don't be shy asking if you'd like help getting oriented to work on something else, and feel free to join our IRC channel #mgmtconfig on Freenode :)

Merged!

**Pull request merged**